server/manager/src/main/java/org/apache/accumulo/manager/tableOps/tableImport/PopulateMetadataTable.java

```diff
@@ -155,11 +166,7 @@ public Repo<Manager> call(long tid, Manager manager) throws Exception {
             if (m == null || !currentRow.equals(metadataRow)) {
 
               if (m != null) {
-                if (!sawTabletAvailability) {
-                  // add a default tablet availability
-                  TabletColumnFamily.AVAILABILITY_COLUMN.put(m,
-                      TabletAvailabilityUtil.toValue(TabletAvailability.ONDEMAND));
-                }
+                AVAILABILITY_COLUMN.put(m, TabletAvailabilityUtil.toValue(initialAvailability));
```

Comment:
The code used to track if it saw a tablet availability.  Now it always overwrites the tablet availability and does not consider if the imported tablets already had availability set.  Need to decide what we want in the following cases.

Case 1 :

 1. User exports a table with tablets having different availability values set.
 2. User imports the table and does not explicitly set a desired availability at import time.

Case 2:

 1.  User exports a table with tablets having different availability values set.
 2.  User imports the table and explicitly sets a desired availability at import time using the new API in this PR.

Case 3:

 1. User exports a table from an older version of Accumulo that does not have tablet availability.
 2. User imports the table and does not explicitly set a desired availability at import time.


For Case 1 I think it should carry forward the availability values from the exported table on import.  For case 2 I think it should override the  the availability values from the exported table on import using the values specified in the API.  For case 3 it should set ONDEMAND for all imported tablets because they are not set in the export data or in the API.

Reply:
>Thats originally why I had null as the default for the tablet availability so that if an override is not provided, we would be able to tell and try to use what is present in the exported table instead. Now that null has been removed, I'm not too sure how we can use what is set in the export if a tablet availability override is not provided.

I we decide we want a certain behavior and null is needed in the implementation to support that behavior then using null seems fine with me.  Sometime optional can be used, but its not always the right choice sometimes trying to use optional becomes extremely cumbersome.

Could try to get the expected behavior written up in the javadoc  for the new method and then do what is needed to implement what is documented.

Reply:
> Now it always overwrites the tablet availability and does not consider if the imported tablets already had availability set.

I had a simiilar thought when reviewing the code. I figured that because this was import table, that it was creating a new table, not importing files into an existing table or tablets. Am I wrong here?

Reply:
>I had a simiilar thought when reviewing the code. I figured that because this was import table, that it was creating a new table, not importing files into an existing table or tablets. Am I wrong here?

Its creating a new table using tablet metadata from an exported table.  The tablet metadata from the exported table may have availability set. We need to decide if we want to ignore this information or use it during import when it creates the new tablets.

Reply:
It may be just requiring the goal on import and ignoring (or never persisting) the goal on export would be the "least surprising" to users.  Maybe the import could default to unhosted.  The user should know at import the desired availability state.  That could have no relationship to what it was when it was exported.  

If we do allow the export to define the availability, then maybe we need a way to allow that state to be examined?  Basically, other than a manual process, how would a user know and track how a table was exported? Does it really matter?

Reply:
Wouldn't the simplest thing to do be:
1. Export the tablet availability
2. Import the tablet availability
3. Have the user use the shell command to change it the way that they want it?

Reply:
Adding to my comment above, IIRC the changes in this PR are an all or nothing approach. It doesn't support the ranges like the shell command. I'm not sure why we are doing this TBH. We are creating something for which a better implementation exists. IMO, we have a solution, we should let someone ask for an enhancement with some actual feedback.

Reply:
@EdColeman - Thanks for the explanation, that does help me understand the concern here. Maybe the easiest solution would be import ONDEMAND and overwrite any existing availability set on the tablet in the export. As you point out, what was HOSTED in the source system may be an issue in the destination system. Then the user can use the shell command post-import to configure it how they want.

Reply:
> Basically, other than a manual process, how would a user know and track how a table was exported? Does it really matter?

Export/import table will transfer tablet metadata like split and tablet configuration in addition to the tables data.  Transferring the tablet availability a user set feels consistent with this existing functionality of transferring table metadata.

As for inspecting a table, the current API allows importing a table and having its initial state be offline.  Could extend what @dlmarion  proposed to be following.

 1. Export  the tablet availability
 2. Import the tablet availability AND set the initial table state to offline
 3. Have the user use the shell command to change it the way that they want it
 4. Online table

Not sure if we can set table availability when a table is offline though, if not then step 3 above would not work.  Was thinking that adding the extra step  of offline helps avoid unwanted tablet hosting.

Reply:
I think the distinction will come down to how availability is perceived.  Is it an inherent property of a table, or is it more closely related to the system and environment.  Things like splits, table properties seem closely tied to the table and the data.  It may be that availability does not have that same affinity - it is more dependent on the system environment at any particular point in time.

As an example, say you had historical data that was infrequently used except  for something like a month or yearly report - so when it comes time to run the report(s) the availability is set to HOSTED while the reports are running, otherwise ONDEMAND is fine for an occasional query.  With that type of usage, the availability is not tied to the table / data, but instead relates directly to the system usage at that point in time.

Reply:
> I think the distinction will come down to how availability is perceived.

It's defined as Tablet Availability.

>  With that type of usage, the availability is not tied to the table / data, but instead relates directly to the system usage at that point in time.

It's a tablet level setting. A tablet with an ONDEMAND tablet availability may be hosted, or not, depending on the usage at that point in time. But it's availability setting is ONDEMAND.

Reply:
Reading back over the comments and thinking about out options, identified the following four possible options.

 1. On table import always set all tablets to ondemand.  No API changes.
 2. On table import use use availability from exported table if it exists, otherwise use ondemand. No API changes.
 3. Add a new API for table import that allows setting availability.  The impl of this API will ignore availability from exported table and only use what was set in the API or ondemand if not set.
 4. Add a new API for table import that allows setting availability.  The impl of this API will use availability from exported table if it exist, otherwise it will use API setting, otherwise it will use ondemand.

I think I am leaning twoards options 1 because its the one where its easiest to understand what will happen when a table is imported.  If we went with option 1 could do the following in this PR.

 * Make this PR just always set ondemand for all new tablets it creates.
 * Update table import javadoc to explain what happens with tablet availability.

If there is a need in the future to maintain availability information, then that could be added with a feature that preserves it on export.

Reply:
Additionally, for 3 and 4, I think the API would need to support setting the availability at the range level. Otherwise the user would have to set something for the import, then log into the shell and make the necessary changes.

Reply:
> Additionally, for 3 and 4, I think the API would need to support setting the availability at the range level. 

That API capability  would need to be supported in the shell also, which would not be nice.

Reply:
> I think the distinction will come down to how availability is perceived. Is it an inherent property of a table, or is it more closely related to the system and environment. Things like splits, table properties seem closely tied to the table and the data. It may be that availability does not have that same affinity - it is more dependent on the system environment at any particular point in time.

It definitely fits into more of a runtime/execution aspect of the tablets metadata like table location, external compaction ids, scan refs, etc.  Where other parts of the tablet metadata are more related to the table data like the range, file list, wals, next tablet timestamp, etc

Reply:
There was a lot of discussion in this thread, so its hard to know what peoples preference is.  Just to see if we are at consensus it may be useful to state your preference.  My preference based on discussion here is : "On table import always set all tablets to ondemand. No API changes.".  @dlmarion, @DomGarguilo , @EdColeman do you have a prefrence?

Reply:
I do not have a strong preference and am willing to defer to others who might.

Reply:
My preference:

Always set all tables to ondemand on import - unless the import command allows the user to specifically set a different, desired state.  If the user knows what they want, it would be nice if they did not have to go through two steps to set the goal.  It should be a uniform state though, all ondemand or all something else. If they want specific ranges, that might require two steps, (import, set-state).  The default, unless specified otherwise by the user should be ondemand.

The state should not be included in the export data.  If it must be, it should be ignored on import.

Reply:
From more discussion it seems like, at least as a first step, all imported tables should be brought up with the on-demand availability. I will work to either make that happen in this PR or close this one in favor of a new PR that does that.